### PR TITLE
Fixes #253 Catch and save standardization and registration errors

### LINKED ISSUE #253: Store restandardization dry-run failures

### PR DESCRIPTION
**Pull request by brianbolt**
I cut a release/1.13.7 branch.

### CODE REVIEW

File: src/main/java/com/labsynch/labseer/domain/StandardizationDryRunCompound.java [hunk 1]
```diff
@@ -225,6 +238,24 @@ private static Predicate buildPredicatesForSearch(CriteriaBuilder criteriaBuilde
 			predicates.add(criteriaBuilder.equal(root.get("asDrawnDisplayChange"), dryRunSearch.getAsDrawnDisplayChange()));
 		}
 
+		// Standardization statuses
+		if (dryRunSearch.getStandardizationStatuses() != null) {
+			if(dryRunSearch.getStandardizationStatuses().length > 0) {
+				predicates.add(root.get("standardizationStatus").in(dryRunSearch.getStandardizationStatuses()));
+			} else {
+				predicates.add(criteriaBuilder.equal(root.get("id"), -1));
```
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/domain/StandardizationDryRunCompound.java:246]: This seems rather aggressive - if no statuses are selected we return zero results, rather than simply not applying a filter? Maybe this will make more sense when I see the UI.

File: src/main/java/com/labsynch/labseer/service/StandardizationServiceImpl.java [hunk 2]
```diff
@@ -395,88 +411,88 @@ public int dupeCheckStandardizationStructures() throws CmpdRegMolFormatException
...
+				if (!newDuplicateCorpNames.equals("")) {
+					dryRunCompound.setNewDuplicates(newDuplicateCorpNames);
+				}
+				dryRunCompound.setExistingDuplicateCount(oldDuplicateCount);
+				if (!oldDuplicateCorpNames.equals("")) {
+					dryRunCompound.setExistingDuplicates(oldDuplicateCorpNames);
 					}
 
-				}
+				// }
```
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/StandardizationServiceImpl.java:495]: what's with this commented out closing brace? Let's remove it

File: src/main/java/com/labsynch/labseer/service/StandardizationServiceImpl.java [hunk 3]
```diff
@@ -395,88 +411,88 @@ public int dupeCheckStandardizationStructures() throws CmpdRegMolFormatException
...
-					hits = chemStructureService.searchMolStructures(cmpdRegMolecules.get(tmpStructureKey),
-							StructureType.DRY_RUN, SearchType.DUPLICATE_TAUTOMER, -1F, -1);
-					newDupeCount = hits.length;
-					for (int hit : hits) {
-						List<StandardizationDryRunCompound> searchResults = StandardizationDryRunCompound
-								.findStandardizationDryRunCompoundsByCdId(hit).getResultList();
-						for (StandardizationDryRunCompound searchResult : searchResults) {
-							if (searchResult.getCorpName().equalsIgnoreCase(dryRunCompound.getCorpName())) {
-								newDupeCount = newDupeCount - 1;
+				// if (dryRunCompound.getNewMolWeight() == 0) {
```
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/StandardizationServiceImpl.java:414]: Let's remove this dead code

File: src/main/java/com/labsynch/labseer/chemclasses/bbchem/ChemStructureServiceBBChemImpl.java [hunk 4]
```diff
@@ -609,22 +621,48 @@ public StrippedSaltDTO stripSalts(CmpdRegMolecule inputStructure) throws CmpdReg
...
-		HashMap<String, String> preprocessedStructures = new HashMap<String, String>();
+		HashMap<String, Entry<String, String>>preprocessedStructures = new HashMap<String, Entry<String, String>>();
 		try {
 			preprocessedStructures = bbChemStructureService.getPreprocessedStructures(structures);
 		} catch (IOException e) {
 			logger.error("Error getting preprocessed structures: "+e.getMessage());
 			throw new CmpdRegMolFormatException("Error getting preprocessed structures: "+e.getMessage());
 		}
 
+		// Only structures which have passed preprocessing should be passed to be processed so we need to remove those with an error
```
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/chemclasses/bbchem/ChemStructureServiceBBChemImpl.java:632]: This comment is a little misleading. We are not "removing" those with an error. We are substituting their original structure in lieu of the preprocessed structure (because we are missing a preprocessed structure). Could call this "filling in preprocessor failures with their original structure"

File: src/main/java/com/labsynch/labseer/chemclasses/bbchem/ChemStructureServiceBBChemImpl.java [hunk 5]
```diff
@@ -609,22 +621,48 @@ public StrippedSaltDTO stripSalts(CmpdRegMolecule inputStructure) throws CmpdReg
...
+		// Loop through the preprocessStructures and remove any structures that have an error
+		HashMap<String, String> structuresToProcess = new HashMap<String, String>();
+		for (String structureKey : preprocessedStructures.keySet()){
+			Entry<String, String> preprocessedStructure = preprocessedStructures.get(structureKey);
+			if (!preprocessedStructure.getKey().equals("SUCCESS")) {
+				logger.error("Got unexpected status '"+preprocessedStructure.getKey()+"' when preprocessing structure for id:'"+structureKey+"' molstruture: '" + structures.get(structureKey) + "'");
+				// Just add the original structure to the list of structures to process
+				// We will add the error message to the structure later
+				// TODO: Make sure the statement ends up true before merging this code
```
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/chemclasses/bbchem/ChemStructureServiceBBChemImpl.java:641]: I think you're doing this down on line 660, right? So can probably remove this TODO

File: src/main/java/com/labsynch/labseer/chemclasses/bbchem/CmpdRegMoleculeBBChemImpl.java [hunk 6]
```diff
@@ -23,8 +23,9 @@
 
 	private BBChemStructureService bbChemStructureService;
 
-	public CmpdRegMoleculeBBChemImpl(BBChemParentStructure molecule) {
+	public CmpdRegMoleculeBBChemImpl(BBChemParentStructure molecule, BBChemStructureService bbChemStructureService) {
 		this.molecule = molecule;
+		this.bbChemStructureService = bbChemStructureService;
```
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/chemclasses/bbchem/CmpdRegMoleculeBBChemImpl.java:28]: Why is this needed?

File: src/main/java/com/labsynch/labseer/chemclasses/bbchem/CmpdRegMoleculeBBChemImpl.java [hunk 7]
```diff
@@ -75,7 +76,8 @@ public String getMolStructure() throws CmpdRegMolFormatException {
 
 	@Override
 	public String getFormula() {
-		if(this.molecule.getMolecularFormula() != null) {
+		if(this.getRegistrationStatus() == RegistrationStatus.ERROR) return null;
+		if(this.molecule.getRegistrationStatus() != RegistrationStatus.ERROR && this.molecule.getMolecularFormula() != null) {
```
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/chemclasses/bbchem/CmpdRegMoleculeBBChemImpl.java:80]: nit: `else if` would make more sense here

**Review by bffrost — CHANGES_REQUESTED** [review]
Mostly looks good. Have a couple requests for cleanup of dead code in some already complicated functions to improve readability.

File: src/main/java/com/labsynch/labseer/chemclasses/bbchem/CmpdRegMoleculeBBChemImpl.java (hunk 6 above)
Comment (brianbolt) [reply src/main/java/com/labsynch/labseer/chemclasses/bbchem/CmpdRegMoleculeBBChemImpl.java:28]: I couldn't get the @autowired annotation to work on this class so anywhere I knew up a CmpdRegMoleculeBBChemImpl I need to pass in the service or it ends up as null.  When debugging this new code, I found that if the CmpdRegMoleculeBBChemImpl class was newed up from a string, then I could save a molecule as it through a null pointer when calling bbchemStructureService.  I can't recall the specific issue in this pull requested that prompted this fix but it was important...

File: src/main/java/com/labsynch/labseer/domain/StandardizationDryRunCompound.java (hunk 1 above)
Comment (brianbolt) [reply src/main/java/com/labsynch/labseer/domain/StandardizationDryRunCompound.java:246]: yeah, I suppose it is agressive and it was questionable when I did it.  My logic

 - Passing in an actual list -> Filter by list
 - Pass null -> Don't filter
 - Pass empty list ->  Return nothing because somebody might create a UI where this is a state of the UI and expect 0 results

File: src/main/java/com/labsynch/labseer/chemclasses/bbchem/ChemStructureServiceBBChemImpl.java (hunk 5 above)
Comment (brianbolt) [reply src/main/java/com/labsynch/labseer/chemclasses/bbchem/ChemStructureServiceBBChemImpl.java]: Yes, thanks forgot to remove this after fixing that.

File: src/main/java/com/labsynch/labseer/domain/StandardizationDryRunCompound.java (hunk 1 above)
Comment (brianbolt) [reply src/main/java/com/labsynch/labseer/domain/StandardizationDryRunCompound.java:246]: FWIW, I did try do a in an empty list but it wasn't working for me so I did this instead.

Comment (bffrost) [reply src/main/java/com/labsynch/labseer/domain/StandardizationDryRunCompound.java:246]: OK. This will go through functional testing so I'm alright with this.

File: src/main/java/com/labsynch/labseer/chemclasses/bbchem/CmpdRegMoleculeBBChemImpl.java [hunk 8]
```diff
@@ -76,8 +76,9 @@ public String getMolStructure() throws CmpdRegMolFormatException {
 
 	@Override
 	public String getFormula() {
-		if(this.getRegistrationStatus() == RegistrationStatus.ERROR) return null;
-		if(this.molecule.getRegistrationStatus() != RegistrationStatus.ERROR && this.molecule.getMolecularFormula() != null) {
+		if(this.getRegistrationStatus() == RegistrationStatus.ERROR) { 
+			return null; 
+		} else if(this.molecule.getRegistrationStatus() != RegistrationStatus.ERROR && this.molecule.getMolecularFormula() != null) {
```
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/chemclasses/bbchem/CmpdRegMoleculeBBChemImpl.java:81]: `this.molecule.getRegistrationStatus() != RegistrationStatus.ERROR &&` is redundant. This can just be
`else if(this.molecule.getMolecularFormula() != null`

**Review by bffrost — APPROVED** [review]
tiny nit, then LGTM

File: src/main/java/com/labsynch/labseer/chemclasses/bbchem/CmpdRegMoleculeBBChemImpl.java (hunk 8 above)
Comment (brianbolt) [reply src/main/java/com/labsynch/labseer/chemclasses/bbchem/CmpdRegMoleculeBBChemImpl.java:81]: Thanks, missed this just pushed. I'll remerge this branch as I jumped the gun

Comment (brianbolt) [reply src/main/java/com/labsynch/labseer/chemclasses/bbchem/CmpdRegMoleculeBBChemImpl.java:81]: Github doesn't let you re-open a pull request so I just just created a new one and merged it: https://github.com/mcneilco/acas-roo-server/pull/259
